src/service/fixed-layer.js

```diff
@@ -0,0 +1,417 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {onDocumentReady} from '../document-state';
+import {log} from '../log';
+import {setStyle, setStyles} from '../style';
+
+const TAG = 'FixedLayer';
+
+
+/**
+ * The fixed layer is a *sibling* of the body element. I.e. it's a direct
+ * child of documentElement. It's used to manage the `postition:fixed`
+ * elements in iOS-iframe case due to the
+ * https://bugs.webkit.org/show_bug.cgi?id=154399 bug, which is itself
+ * a result of workaround for the issue where scrolling is not supported
+ * in iframes (https://bugs.webkit.org/show_bug.cgi?id=149264).
+ * This implementation finds all elements that could be `fixed` and checks
+ * on major relayouts if they are indeed `fixed`. All `fixed` elements are
+ * moved into the fixed layer.
+ */
+export class FixedLayer {
+  /**
+   * @param {!Document} doc
+   * @param {!Vsync} vsync
+   * @param {number} paddingTop
+   * @param {boolean} transfer
+   */
+  constructor(doc, vsync, paddingTop, transfer) {
+    /** @const {!Document} */
+    this.doc = doc;
+
+    /** @private @const {!Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {number} */
+    this.paddingTop_ = paddingTop;
+
+    /** @private @const {boolean} */
+    this.transfer_ = transfer;
```

Comment:
I meant:

warn to the console if this document would have transferred layers if it was viewed inside of a viewer.